src/libraries/Common/src/System/Collections/Generic/ToArray.SpeedOpt.cs

```diff
@@ -0,0 +1,549 @@
+        private const int maxByteElementsArraySize  = 0X7FFFFFC7;
+        private const int maxOtherElementsArraySize = 0X7FEFFFFF;
+
+        protected const int maxArraySize = maxByteElementsArraySize;
+
+        private const int initialRecursiveDepth = 16;
+        private const int itemsPerRecursion = 4;
+
+        protected const int maxSizeForNoAllocations = initialRecursiveDepth * itemsPerRecursion;
+
+        private int GetBufferSize(int count, int sourceSize)
+        {
+            if (count >= maxByteElementsArraySize)
+            {
+                // We fail here "early" as we could be creating more buffers (which would be smaller thaat the max)
+                // but would eventually lead to an excessive allocation which would throw an OutOfMemory exception
+                throw new OutOfMemoryException();
+            }
+
+            var maxSize =
+                count < sourceSize
+                ? sourceSize
+                :   count < maxOtherElementsArraySize
+                    ? maxOtherElementsArraySize
+                    : maxByteElementsArraySize;
+
+            return (int)Math.Min((uint)maxSize, (uint)count * 2) - count; ;
+        }
+
+        protected T[] FinishViaAllocations(object source, int sourceSize, ref int sourceIdx, Func<T, bool>? predicate, int count)
+        {
+            T[] result;
+            int bufferSize = GetBufferSize(count, sourceSize);
+            T[] buffer = new T[bufferSize];
+
+            var (index, moveNext) = PopulateBuffer(buffer, source, ref sourceIdx, predicate);
+
+            if (moveNext)
+            {
+                result = FinishViaAllocations(source, sourceSize, ref sourceIdx, predicate, count + index);
+            }
+            else
+            {
+                result = new T[count + index];
+            }
+
+            Array.Copy(buffer, 0, result, count, index);
+
+            return result;
+        }
+
+        protected abstract (int, bool) PopulateBuffer(T[] buffer, object source, ref int sourceIdx, Func<T, bool>? predicate);
+
+        protected static T[] Allocate(int count)
+        {
+            if (count == 0)
+                return Array.Empty<T>();
+
+            return new T[count];
+        }
+
+        protected static T[] AllocateAndAssign(int count, T item1)
+        {
+            T[] result = new T[count];
+
+            result[--count] = item1;
+
+            return result;
+        }
+
+        protected static T[] AllocateAndAssign(int count, T item1, T item2)
+        {
+            T[] result = new T[count];
+
+            result[--count] = item2;
+            result[--count] = item1;
+
+            return result;
+        }
+
+        protected static T[] AllocateAndAssign(int count, T item1, T item2, T item3)
+        {
+            T[] result = new T[count];
+
+            result[--count] = item3;
+            result[--count] = item2;
+            result[--count] = item1;
+
+            return result;
```

Comment:
Newbie C# question: could this be written as `return new T[3]{item1, item2, item3}`, and asserting that `count` is indeed `3`?

Reply:
It could if 'count' was always 3 (and you wouldn't need to specify 3 in the array allocation), but this call is from a recursive function and the application can be of different sizes, where the assignment just sets the last 3 elements and then it returns up the call stack the rest of the elements are filled in.